src/main/java/com/artipie/api/RepositoryName.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * The MIT License (MIT) Copyright (c) 2020-2021 artipie.com
+ * https://github.com/artipie/artipie/LICENSE.txt
+ */
+
+package com.artipie.api;
+
+import io.vertx.ext.web.RoutingContext;
+
+/**
+ * Repository name.
+ * @since 0.26
+ */
+public interface RepositoryName {
+    /**
+     * Username path parameter name.
+     */
+    String UNAME = "uname";
+
+    /**
+     * Repository path parameter name.
+     */
+    String RNAME = "rname";
+
+    /**
+     * String representation of repository name.
+     * For flat layout consists of 'reponame'
+     * For org layout consists of 'username/reponame'
+     * @return Repository name as string
+     */
+    String string();
+
+    /**
+     * Decorator based on request and layout.
+     * @since 0.26
+     */
+    @SuppressWarnings("PMD.ConstructorOnlyInitializesOrCallOtherConstructors")
+    class FromRequest implements RepositoryName {
+        /**
+         * Decorated repository name.
+         */
+        private final RepositoryName origin;
+
+        /**
+         * Ctor.
+         * @param ctx RoutingContext
+         * @param layout Layout
+         */
+        public FromRequest(final RoutingContext ctx, final String layout) {
+            if ("flat".equals(layout)) {
+                this.origin = new RepositoryName.FlatRepositoryName(
+                    ctx.pathParam(RepositoryName.RNAME)
+                );
+            } else {
+                this.origin = new RepositoryName.OrgRepositoryName(
+                    ctx.pathParam(RepositoryName.UNAME),
+                    ctx.pathParam(RepositoryName.RNAME)
+                );
+            }
```

Comment:
hm, I'd suggest putting all the logic inside `string()` method, then we will not disrupt `ConstructorOnlyInitializesOrCallOtherConstructors` rule and can get rid of implementations `FlatRepositoryName` and `OrgRepositoryName` as they do nothing...